### PR TITLE
Overhaul API and add cross platform hole punching

### LINKED ISSUE #1: Feature: FSCTL_SET_ZERO_DATA
In the future when hypercore starts needing to do garbage collection I've implemented the windows control codes for hole punching a sparse file with the [`FSCTL_SET_ZERO_DATA`](https://docs.microsoft.com/en-us/windows/win32/api/winioctl/ni-winioctl-fsctl_set_zero_data) IoControl code. 

You can see the implementation here, feel free to use it or ping me to whip up a PR.
https://github.com/xori/set-sparse/blob/a6244dec1f67af92b0d9b627f47ad480f152cab8/src/sparse.cpp#L24-L42

This deallocates the file space on the hard drive between two points if it's large enough.

### PR DESCRIPTION
**Pull request by kasperisager**
Closes #1.

### CODE REVIEW

**Comment by kasperisager** [comment]
@mafintosh This is ready for review.

**Comment by kasperisager** [comment]
@mafintosh This doesn't yet handle block alignment on macOS when hole punching.

File: README.md [hunk 1]
```diff
@@ -1,48 +1,76 @@
...
-// Can we unlock it?
-console.log(unlock(fd))
+  write(fd, 'hello world', async (err) => {
+    if (err) throw err
 
-// Can we set the file as sparse?
-console.log(sparse(fd))
+    await unlock(fd)
+  })
+})
```
Comment (mafintosh) [inline README.md]: isn't this way clearer now if we use fs/promises? or do they hide the fd there?

File: README.md [hunk 2]
```diff
@@ -1,48 +1,76 @@
...
+  await lock(fd, { exclusive: true })
 
-// Can we unlock it?
-console.log(unlock(fd))
+  write(fd, 'hello world', async (err) => {
+    if (err) throw err
 
-// Can we set the file as sparse?
-console.log(sparse(fd))
+    await unlock(fd)
```
Comment (mafintosh) [inline README.md]: unlock is sync

File: README.md [hunk 3]
```diff
@@ -1,48 +1,76 @@
...
+Options are the same as `fsctl.lock()`.
 
-Note that the lock is only advisory and there is nothing stopping someone from accessing the file by simply ignoring the lock.
+#### `fsctl.unlock(fd[, offset[, length]])`
 
-Works across processes as well.
+Release a process level lock on a file.
 
-#### `bool = fsctl.unlock(fd)`
+#### `await fsctl.punchHole(fd, offset, length)`
```
Comment (mafintosh) [inline README.md:61]: is unlinkBlocks a better generic name for this? punchHole is obvs cool also 😎

**Comment by mafintosh** [comment]
Looks amazing!

File: README.md (hunk 3 above)
Comment (kasperisager) [reply README.md:61]: Also wondering if we should get rid of `setSparse()` and just do two `DeviceIoControl()` calls when hole punching on Windows. With that in place, would `unlinkBlocks(fd, index, count = 1)` make sense?

Comment (kasperisager) [reply README.md:61]: That would obviously need to do a stat to get the block size first.

Comment (mafintosh) [reply README.md:61]: Block size is basically fixed but yea, we need setSparse to initially set files to sparse on win unless I misunderstood what you say?

Comment (mafintosh) [reply README.md:61]: Maybe what we have is fine…

Comment (kasperisager) [reply README.md:61]: My idea was to simply mark the file as sparse directly in `fsctl__punch_hole()` on Windows. Regarding block sizes, I don't know how often people configure their disks with non-default block sizes. A brief search around the internet revealed that it at least seems to happen when folks deal with massive disks, i.e. several terabytes, where larger block sizes are needed.

Comment (mafintosh) [reply README.md:61]: Ah ok, fair. So you'll still keep the setSparse api yea?

Comment (mafintosh) [reply README.md:61]: Does the underlying syscall use blocks or bytes as offsets/length? We should just mimick that I guess if the same across OSs

Comment (kasperisager) [reply README.md:61]: My suggestion would be to remove it and just do this on Windows:

```c
int
fsctl__punch_hole (uv_os_fd_t fd, uint64_t offset, size_t length) {
  FILE_ZERO_DATA_INFORMATION data = {
    .FileOffset = {
      .QuadPart = offset,
    },
    .BeyondFinalZero = {
      .QuadPart = offset + length,
    },
  };

  DWORD bytes;

  BOOL res = DeviceIoControl(
    fd,
    FSCTL_SET_SPARSE,
    NULL,
    0,
    NULL,
    0,
    &bytes, // Must be passed when lpOverlapped is NULL
    NULL
  );

  if (!res) return uv_translate_sys_error(GetLastError());

  res = DeviceIoControl(
    fd,
    FSCTL_SET_ZERO_DATA,
    &data,
    sizeof(data),
    NULL,
    0,
    &bytes, // Must be passed when lpOverlapped is NULL
    NULL
  );

  return res ? 0 : uv_translate_sys_error(GetLastError());
}
```

Comment (mafintosh) [reply README.md:61]: If we remove setSparse how do we mark new files as sparse? We need that for our data structures as well :)

Comment (kasperisager) [reply README.md:61]: Great point, I'm likely reading the Windows docs wrong! This is the bit that caught my eye:

> When you perform a write operation (with a function or operation other than FSCTL_SET_ZERO_DATA) whose data consists of nothing but zeros, zeros will be written to the disk for the entire length of the write. To zero out a range of the file and maintain sparseness, use FSCTL_SET_ZERO_DATA.

https://docs.microsoft.com/en-us/windows/win32/fileio/sparse-file-operations

I was therefore under the impression that sparseness wouldn't be maintained for normal writes like on non-Windows systems. Let me test that to make sure I got it wrong.

Regarding the hole punching APIs, they use byte offset and length on all systems that support them.

Comment (kasperisager) [reply README.md:61]: Windows behaves like macOS and Linux on sparse writes so all is well and we definitely need `setSparse()` 👌

Comment (kasperisager) [reply README.md:61]: Should we change the name?

Comment (mafintosh) [reply README.md:61]: Let's keep it as is, I overthought it.

**Pull request merged**